### PR TITLE
[SMALLFIX] Use InstancedConfiguration in UfsConfiguration

### PR DESCRIPTION
**Pull request by madanadit**

### CODE REVIEW

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java [hunk 1]
```diff
@@ -127,7 +131,7 @@ public UnderFileSystemConfiguration setShared(boolean shared) {
    * @return the updated configuration object
    */
   public UnderFileSystemConfiguration setUserSpecifiedConf(Map<String, String> ufsConf) {
-    mUfsConf = ufsConf;
+    mUfsConf.merge(ufsConf, Source.UFS_OPTION);
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: this is a little different from before - it might leave behind some key/value pairs from the previous conf. Is this intentional?

**Comment by madanadit** [comment]
@apc999 @aaudiber Please take a look. Every instance of `UfsConfig` now has a copy of all properties from `Configuration`.

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java [hunk 2]
```diff
@@ -62,31 +71,30 @@ public boolean containsKey(PropertyKey key) {
...
-    throw new RuntimeException("key " + key + " not found");
+    return mUfsConf.get(key);
   }
 
   /**
-   * @return the map of user-customized configuration
+   * @return the map of resolved user-customized configuration
    */
   public Map<String, String> getUserSpecifiedConf() {
     if (mUfsConf == null) {
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: is this still needed?

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java [hunk 3]
```diff
@@ -62,31 +71,30 @@ public boolean containsKey(PropertyKey key) {
...
       return Collections.emptyMap();
     }
-    return Collections.unmodifiableMap(mUfsConf);
+    Map<String, String> map = new HashMap<>();
+    mUfsConf.keySet().forEach(key -> {
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: optional: could be clearer to write this as a stream + filter + map

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java [hunk 4]
```diff
@@ -62,31 +71,30 @@ public boolean containsKey(PropertyKey key) {
...
+    return map;
   }
 
   /**
    * @return all the global and mount specific properties as an immutable map
    */
   public Map<String, String> toMap() {
-    Map<String, String> all = new HashMap<>(Configuration.toMap());
-    all.putAll(mUfsConf);
+    Map<String, String> all = new HashMap<>(mUfsConf.toMap());
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: might as well inline this

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java [hunk 5]
```diff
@@ -26,7 +29,8 @@
 public final class UnderFileSystemConfiguration {
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: update class-level javadoc

File: core/common/src/main/java/alluxio/conf/Source.java [hunk 6]
```diff
@@ -52,6 +53,10 @@
      * HadoopConf). This source has the highest priority.
      */
     RUNTIME,
+    /**
+     * The property value is specified as options for a mount point.
+     */
+    UFS_OPTION,
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/conf/Source.java]: perhaps call this `MOUNT_OPTION`?

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java [hunk 7]
```diff
@@ -62,31 +71,30 @@ public boolean containsKey(PropertyKey key) {
...
-    if (Configuration.isSet(key)) {
-      return Configuration.get(key);
-    }
-    throw new RuntimeException("key " + key + " not found");
+    return mUfsConf.get(key);
   }
 
   /**
-   * @return the map of user-customized configuration
+   * @return the map of resolved user-customized configuration
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: Consider replacing "user-customized" and "UserSpecified" with ufs-specific

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java [hunk 8]
```diff
@@ -26,7 +29,8 @@
 public final class UnderFileSystemConfiguration {
   private boolean mReadOnly;
   private boolean mShared;
-  private Map<String, String> mUfsConf;
+  private AlluxioProperties mProperties;
```
[the remark this answers is not in the document]
Comment (madanadit) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: got it

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java (hunk 2 above)
Comment (madanadit) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: nope

File: core/common/src/main/java/alluxio/conf/Source.java [hunk 9]
```diff
@@ -52,6 +53,10 @@
      * HadoopConf). This source has the highest priority.
      */
     RUNTIME,
+    /**
+     * The property value is specified as options for a mount point.
+     */
+    MOUNT_OPTION,
```
Comment (apc999) [inline core/common/src/main/java/alluxio/conf/Source.java]: are we intentional to have mount option with higher priority than the runtime ?

Comment (madanadit) [reply core/common/src/main/java/alluxio/conf/Source.java]: This is consistent with the present behavior. I think we should keep it that way. I’ll update the comment above.

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java [hunk 10]
```diff
@@ -41,16 +56,15 @@ public static UnderFileSystemConfiguration defaults() {
...
     mShared = false;
-    mUfsConf = Collections.EMPTY_MAP;
+    mUfsConf = new InstancedConfiguration(Configuration.copyProperties());
   }
 
   /**
    * @param key property key
    * @return true if the key is contained in the given UFS configuration or global configuration
    */
   public boolean containsKey(PropertyKey key) {
```
Comment (apc999) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: shall we change this to `isSet` too? Fine to leave this in a separate PR to minimize the footprint

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java [hunk 11]
```diff
@@ -62,32 +76,27 @@ public boolean containsKey(PropertyKey key) {
    * @return the value associated with the given key
    */
   public String getValue(PropertyKey key) {
```
Comment (apc999) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: maybe better to be `get` too?

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java [hunk 12]
```diff
@@ -62,32 +76,27 @@ public boolean containsKey(PropertyKey key) {
...
   /**
-   * @return the map of user-customized configuration
+   * @return the map of resolved ufs specific configuration
    */
-  public Map<String, String> getUserSpecifiedConf() {
-    if (mUfsConf == null) {
-      return Collections.emptyMap();
-    }
-    return Collections.unmodifiableMap(mUfsConf);
+  public Map<String, String> getUfsSpecificConf() {
```
Comment (apc999) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: is it more proper to be `getMountSpecificConf` ? up to you

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java (hunk 5 above)
Comment (apc999) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: should this just be a subclass of `InstancedConfiguration`, with additional `readonly` `shared` and `ufsConf` fields?

File: core/common/src/main/java/alluxio/conf/Source.java (hunk 9 above)
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/Source.java]: make sense

**Comment by apc999** [comment]
question: should the `UnderFileSystemConfiguration` just be a subclass of `InstancedConfiguration`, with additional getters and setters for `readonly` `shared` and `ufsConf` fields? 
otherwise LGTM

**Comment by madanadit** [comment]
@aaudiber @apc999 Please take another look

File: core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java [hunk 13]
```diff
@@ -12,21 +12,35 @@
...
+ * </ol>
+ *
+ * <p>
+ * This class extends {@link InstancedConfiguration}. Variable substitution and aliases
+ * are supported.
  */
-public final class UnderFileSystemConfiguration {
+@NotThreadSafe
+@PublicApi
+public final class UnderFileSystemConfiguration extends InstancedConfiguration {
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: This creates a tight coupling between `UnderFileSystemConfiguration` and `InstancedConfiguration`, where any internal change to `InstancedConfiguration` needs to worry about breaking `UnderFileSystemConfiguration`, and `UnderFileSystemConfiguration` needs to be aware of the implementation details of `InstancedConfiguration`.

To avoid the tight coupling I suggest implementing `AlluxioConfiguration` instead, and delegating to an internal instance of `InstancedConfiguration`. Then this class depends only on the public API of `InstancedConfiguration`. It will add boilerplate to delegate each method, but I think it is worth it to keep the classes isolated.

Comment (madanadit) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: since we are going to re-iterate the configuration classes later, i vote to keep it the way it is for now and defer further changes to a follow-up PR. let me know if you guys are good with that

Comment (aaudiber) [reply core/common/src/main/java/alluxio/underfs/UnderFileSystemConfiguration.java]: sgtm
